torchdata/datapipes/iter/load/iopath.py

```diff
@@ -183,8 +184,11 @@ def __init__(
     def __iter__(self) -> Iterator[str]:
         for meta, data in self.source_datapipe:
             filepath = meta if self.filepath_fn is None else self.filepath_fn(meta)
-            with self.pathmgr.open(filepath, self.mode) as f:
-                f.write(data)
+
+            if not os.path.exists(filepath):
+                with self.pathmgr.file_lock(filepath):
```

Comment:
Two questions:
1. Not familiar with this `if not os.path.exists(filepath)`. Does this means it only tries to lock if the path doesn't exist?
2. What is the behavior here if the file is already open elsewhere (perhaps there is another `file_lock`)? Does it perform a non-blocking wait?

Reply:
This was in reference to https://www.internalfb.com/diff/D35459528?dst_version_fbid=2351551101810096&transaction_fbid=471804224682410

Ask was to skip writing to file if it already exists so for now I brought that check down to the write

Moved to PR to draft still scoping out the problem and solution, just wanted a way to communicate with you and @ejguan

Reply:
> Not familiar with this if not os.path.exists(filepath). Does this means it only tries to lock if the path doesn't exist?

I believe `filelock` will create a lock file as a mutex across processes. It's not using `os.path.exists(filepath)` to lock.

The implementation is a little bit weird though. I think we should do:
```py
            with self.pathmgr.file_lock(filepath):
                if not os.path.exists(filepath):
                    with self.pathmgr.open(filepath, self.mode) as f:
                        f.write(data)
```